From 402edc71afb0c03ad7bf046b5d0abcdd8e0497df Mon Sep 17 00:00:00 2001
From: Matthias Clasen <mclasen@redhat.com>
Date: Tue, 22 Feb 2022 19:48:49 -0500
Subject: [PATCH] Fix rendering of color glyphs on a clear surface

We were calling the backend vfuncs directly, thereby
missing some of the surface state bookkeeping that
is done in the wrappers.

Fixes: #539
---
 src/cairo-surface.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/src/cairo-surface.c b/src/cairo-surface.c
index 46f39cb20..018436350 100644
--- a/src/cairo-surface.c
+++ b/src/cairo-surface.c
@@ -2667,10 +2667,10 @@ composite_one_color_glyph (cairo_surface_t       *surface,
         pattern = cairo_pattern_create_for_surface ((cairo_surface_t *)glyph_surface);
         cairo_matrix_init_translate (&matrix, - x, - y);
         cairo_pattern_set_matrix (pattern, &matrix);
-	if (op == CAIRO_OPERATOR_SOURCE || op == CAIRO_OPERATOR_CLEAR || !has_color)
-	  status = surface->backend->mask (surface, op, pattern, pattern, clip);
-	else
-	  status = surface->backend->paint (surface, op, pattern, clip);
+        if (op == CAIRO_OPERATOR_SOURCE || op == CAIRO_OPERATOR_CLEAR || !has_color)
+          status = _cairo_surface_mask (surface, op, pattern, pattern, clip);
+        else
+          status = _cairo_surface_paint (surface, op, pattern, clip);
         cairo_pattern_destroy (pattern);
     }
 
-- 
GitLab

